tests/grid/grid_tools_transform_02.cc

```diff
@@ -0,0 +1,76 @@
+// ------------------------------------------------------------------------
+//
+// SPDX-License-Identifier: LGPL-2.1-or-later
+// Copyright (C) 2001 - 2024 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// Part of the source code is dual licensed under Apache-2.0 WITH
+// LLVM-exception OR LGPL-2.1-or-later. Detailed license information
+// governing the source code and code contributions can be found in
+// LICENSE.md and CONTRIBUTING.md at the top level directory of deal.II.
+//
+// ------------------------------------------------------------------------
+
+
+
+#include <deal.II/grid/grid_generator.h>
+#include <deal.II/grid/grid_out.h>
+#include <deal.II/grid/grid_tools.h>
+#include <deal.II/grid/tria.h>
+
+#include "../tests.h"
+
+
+Point<3>
+trans_func(const Point<3> &p)
+{
+  Point<3> r(p[0] + p[1] * p[1], p[1], p[2]);
+  return r;
+}
+
+
+
+void
+test()
+{
+  Triangulation<3> tria;
+  GridGenerator::hyper_cube(tria);
+  tria.refine_global(1);
+  tria.execute_coarsening_and_refinement();
+
+
+  typename Triangulation<3, 3>::active_cell_iterator cell = tria.begin_active();
+  for (unsigned int i = 0; i < 3; i++)
+    {
+      cell->set_refine_flag(RefinementCase<3>::cut_axis(i));
+      cell++;
+    }
+  tria.execute_coarsening_and_refinement();
+
+  deallog << "Unchanged grid:" << std::endl;
+  GridOut().write_gnuplot(tria, deallog.get_file_stream());
+  {
+    std::ofstream f("grid1");
+    GridOut().write_gnuplot(tria, f);
+  }
+
+  GridTools::transform(trans_func<3>, tria);
+  deallog << "transformed grid:" << std::endl;
+  GridOut().write_gnuplot(tria, deallog.get_file_stream());
+  {
+    std::ofstream f("grid2");
+    GridOut().write_gnuplot(tria, f);
+  }
```

Comment:
How are you verifying that this is correct?

Reply:
I don't. I just check, if it runs through. Isn't that what you do in the other test `tests/grid/grid_tools_transform_01.cc`? I haven't seen the comparison with some output file there. 

I'm sorry, I have never written tests before.

Reply:
Something like
```diff
diff --git a/tests/grid/grid_tools_transform_02.cc b/tests/grid/grid_tools_transform_02.cc
index 58fee473c9..1f3f9d6cf7 100644
--- a/tests/grid/grid_tools_transform_02.cc
+++ b/tests/grid/grid_tools_transform_02.cc
@@ -35,33 +35,65 @@ void
 test()
 {
   Triangulation<3> tria;
-  GridGenerator::hyper_cube(tria);
-  tria.refine_global(1);
-  tria.execute_coarsening_and_refinement();
+  {
+    GridGenerator::hyper_cube(tria);
+    tria.refine_global(1);
+    tria.execute_coarsening_and_refinement();
+
+    typename Triangulation<3, 3>::active_cell_iterator cell =
+      tria.begin_active();
+    for (unsigned int i = 0; i < 3; i++)
+      {
+        cell->set_refine_flag(RefinementCase<3>::cut_axis(i));
+        cell++;
+      }
+    tria.execute_coarsening_and_refinement();
+
+    deallog << "Unchanged grid:" << std::endl;
+    GridOut().write_gnuplot(tria, deallog.get_file_stream());
+    {
+      std::ofstream f("grid1.vtk");
+      GridOut().write_vtk(tria, f);
+    }
+
+    GridTools::transform(trans_func, tria);
+    deallog << "transformed grid:" << std::endl;
+    GridOut().write_gnuplot(tria, deallog.get_file_stream());
+    {
+      std::ofstream f("grid2.vtk");
+      GridOut().write_vtk(tria, f);
+    }
+  }
 
+  Triangulation<3> tria_ref;
+  GridGenerator::hyper_cube(tria_ref);
+  tria_ref.refine_global(1);
+  tria_ref.execute_coarsening_and_refinement();
 
-  typename Triangulation<3, 3>::active_cell_iterator cell = tria.begin_active();
+  typename Triangulation<3, 3>::active_cell_iterator cell =
+    tria_ref.begin_active();
   for (unsigned int i = 0; i < 3; i++)
     {
       cell->set_refine_flag(RefinementCase<3>::cut_axis(i));
       cell++;
     }
-  tria.execute_coarsening_and_refinement();
+  tria_ref.execute_coarsening_and_refinement();
 
-  deallog << "Unchanged grid:" << std::endl;
-  GridOut().write_gnuplot(tria, deallog.get_file_stream());
-  {
-    std::ofstream f("grid1");
-    GridOut().write_gnuplot(tria, f);
-  }
-
-  GridTools::transform(trans_func<3>, tria);
-  deallog << "transformed grid:" << std::endl;
-  GridOut().write_gnuplot(tria, deallog.get_file_stream());
-  {
-    std::ofstream f("grid2");
-    GridOut().write_gnuplot(tria, f);
-  }
+  auto active_cell  = tria.begin_active();
+  auto tria_ref_end = tria_ref.end();
+  for (auto active_cell_ref = tria_ref.begin_active();
+       active_cell_ref != tria_ref_end;
+       ++active_cell_ref, ++active_cell)
+    {
+      for (unsigned int i = 0; i < 8; ++i)
+        {
+          auto transformed_vertex = trans_func(active_cell_ref->vertex(i));
+          std::cout << active_cell->vertex(i) << " should be "
+                    << transformed_vertex << std::endl;
+          Assert(active_cell->vertex(i) == transformed_vertex,
+                 ExcInternalError());
+        }
+    }
 }
```
which fails with
```
[...]
0.5 0.5 0 should be 0.5 0.5 0
0.75 0.5 0 should be 0.75 0.5 0
0.3125 0.25 0.25 should be 0.3125 0.25 0.25
0.5625 0.25 0.25 should be 0.5625 0.25 0.25
0.5 0.5 0.25 should be 0.5 0.5 0.25
0.75 0.5 0.25 should be 0.75 0.5 0.25
0 0 0.25 should be 0 0 0.25
0.25 0 0.25 should be 0.25 0 0.25
0.0625 0.25 0.25 should be 0.0625 0.25 0.25
0.3125 0.25 0.25 should be 0.3125 0.25 0.25
0 0 0.5 should be 0 0 0.5
0.25 0 0.5 should be 0.25 0 0.5
0.125 0.25 0.5 should be 0.0625 0.25 0.5
libc++abi: terminating due to uncaught exception of type dealii::StandardExceptions::ExcInternalError
Abort trap: 6
```

Reply:
The check needs some consideration for hanging nodes, though.

Reply:
It's quite possible that a correct check is just as complicated as the implementation. In that case, we should at least confirm manually that the output is correct.

Reply:
Well, it's no wonder this test fails, due to hanging nodes. That test would also fail for isotropically refined cells, which are locally refined. The reason: Hanging nodes are set to the midpoint of 'its parents line vertices'. I don't know how to put it more elegantly. E.g. consider the hanging node `v` that sits on line `L`. Instead of `trans_func(v)`, we safe 
```
(trans_func(L->vertex(0)) + trans_func(L->vertex(1))/2.0; 
```
which is obviously not the same as `trans_func(v)`, whenever `trans_func()` is some non-linear function (as is here). 

This is actually, what I was wondering about: Why are hanging nodes fixed in `GridTools::transform(...)` the way they are? 
> To fix holes
is what it says right before. But, this code inevitably overwrites the given transformation to achieve that.

Reply:
So, I have tried to test for hanging edges:
```
  std::vector<bool> treated_vertices(tria_ref.n_vertices(), false);
  auto active_cell  = tria.begin_active();
  auto tria_ref_end = tria_ref.end();
  bool success = true;
  for (auto active_cell_ref = tria_ref.begin_active();
            active_cell_ref != tria_ref_end, success;
            ++active_cell_ref, ++active_cell)
    {
      // First, check hanging nodes
      std::cout << "Testing for hanging nodes on " 
                << active_cell_ref->level() << "." 
                << active_cell_ref->index() << std::endl;
      unsigned int failed_on_face;
      for (unsigned int face = 0; face < 6 && success; ++face)
          if (active_cell_ref->face(face)->has_children() && 
                !active_cell_ref->face(face)->at_boundary())
              for (unsigned int line = 0; line < 4 && success; ++line)
                if (active_cell_ref->face(face)->line(line)->has_children())
                  {
                    auto transformed_vertex = (
                            trans_func(
                                active_cell_ref->face(face)->line(line)->vertex(0)
                            ) +
                            trans_func(
                                active_cell_ref->face(face)->line(line)->vertex(1)
                            ) ) / 2.0;
                    std::cout << active_cell->face(face)->line(line)->child(0)->vertex(1)
                              << " should be " << transformed_vertex << std::endl;
                    treated_vertices[active_cell->face(face)->line(line)->child(0)->vertex_index(1)] = true;
                    success = active_cell->face(face)->line(line)->child(0)->vertex(1) == transformed_vertex;
                    if (!success)
                      failed_on_face = face;
                  } 

      // Then, check all the other vertices. This block is chunky, because I wanted to know exactly where the test fails
      std::cout << "Testing remaining vertices on " 
                << active_cell_ref->level() << "." 
                << active_cell_ref->index() << std::endl;
      for (unsigned int face = 0; face < 6 && success; ++face)
        for (unsigned int line = 0; line < 4 && success; ++line)
          for (unsigned int vertex = 0; vertex < 2 && success; ++vertex) 
            if(!treated_vertices[active_cell->face(face)->line(line)->vertex_index(vertex)])
              {
                auto transformed_vertex = 
                      trans_func(active_cell_ref->face(face)->line(line)->vertex(vertex));
                std::cout << active_cell->face(face)->line(line)->vertex(vertex) << " should be "
                          << transformed_vertex << std::endl;
                treated_vertices[active_cell->face(face)->line(line)->vertex_index(vertex)] = true;
                success = active_cell->face(face)->line(line)->vertex(vertex) == transformed_vertex;
                if (!success)
                  failed_on_face = face;
              }

      // Collect information about failure
      if (!success){
        std::cout << "Refinement cases: " << std::endl;
        std::cout << unsigned(RefinementCase<3>::isotropic_refinement) << ": isotropic" << std::endl;
        std::cout << unsigned(RefinementCase<3>::no_refinement) << ": none" << std::endl;
        std::cout << unsigned(RefinementCase<3>::cut_axis(0)) << ": x-axis refined" << std::endl;
        std::cout << unsigned(RefinementCase<3>::cut_axis(1)) << ": y-axis refined" << std::endl;
        std::cout << unsigned(RefinementCase<3>::cut_axis(2)) << ": z-axis refined" << std::endl;

        std::cout << "Failed cell: " << active_cell->level() << "." << active_cell->index() << std::endl;
        std::cout << "Neighbor: ";
        if (active_cell->neighbor(failed_on_face).state() == IteratorState::valid)
          std::cout << active_cell->neighbor(failed_on_face)->level() << "."
                    << active_cell->neighbor(failed_on_face)->index() << std::endl;
        else 
          std::cout << "none" << std::endl;

        std::cout << "Parent: ";
        if (active_cell->level() > 0)
          std::cout << active_cell->parent()->level() << "." 
                    << active_cell->parent()->index() << std::endl
                    << "RefinementCase = " 
                    << unsigned(static_cast<uint8_t>(active_cell->parent()->refinement_case()))
                    << std::endl;
        else 
          std::cout << "none" << std::endl;
      }

      Assert(success, ExcInternalError());
    }
```
With the following output 

```
...
0.375 0.25 0.5 should be 0.3125 0.25 0.5
Refinement cases: 
7: isotropic
0: none
1: x-axis refined
2: y-axis refined
4: z-axis refined
Failed cell: 2.4
Neighbor: 2.5
Parent: 1.0
RefinementCase = 7
```
And now I'm wondering as to why the refinement case of cell `1.0` has been changed from `RefinementCase<3>::cut_axis(0)` (aka x-refinement) to `RefinementCase<3>::isotropic_refinement`. This makes it hard for me to understand why this fails.

Reply:
I should add: The output in`.vtk` is also quite strange; there are lines missing.